spec/integration/data_stream/manifest.spec.yml

```diff
@@ -307,6 +332,29 @@ spec:
         dynamic_namespace:
           description: When set to true, agents running this integration are granted data stream privileges for all namespaces of its type
           type: boolean
+    routing_rules:  # technical preview
+      description: Routing rules set.
+      type: array
+      items:
```

Comment:
Applying this suggestion:
```diff
--- spec/integration/data_stream/manifest.spec.yml
+++ spec/integration/data_stream/manifest.spec.yml
@@ -358,20 +358,11 @@ spec:
           type: boolean
     routing_rules:  # technical preview
       description: Routing rules set.
-      type: array
-      items:
-        type: object
-        additionalProperties: false
-        properties:
-          dataset:
-            type: string
-          rules:
-            type: array
-            items:
-              $ref: "#/definitions/routing_rule"
-        required:
-          - dataset
-          - rules
+      type: object
+      additionalProperties:
+        type: array
+        items:
+          $ref: "#/definitions/routing_rule"
   allOf:
     - if:
         required:
```

works and it would be like in the example given in the issue.

But it cannot be applied, since some elements under `routing_rules` could contain dots, for instance `k8s.router`. As it is implemented now package-spec, every dot in these keys are expanded as objects ([link](https://github.com/elastic/package-spec/blob/5e86c23605e14f5800a966a7ab535ef106b5f0b8/code/go/internal/yamlschema/loader.go#L92))

I think we would need to keep this a list of elements